cpp/include/cuvs/neighbors/vamana.hpp

```diff
@@ -0,0 +1,326 @@
+    return data_rows > 0 ? data_rows : graph_view_.extent(0);
+  }
+
+  /** Dimensionality of the data. */
+  [[nodiscard]] constexpr inline auto dim() const noexcept -> uint32_t { return dataset_->dim(); }
+  /** Graph degree */
+  [[nodiscard]] constexpr inline auto graph_degree() const noexcept -> uint32_t
+  {
+    return graph_view_.extent(1);
+  }
+
+  [[nodiscard]] inline auto dataset() const noexcept
+    -> raft::device_matrix_view<const T, int64_t, raft::layout_stride>
+  {
+    auto p = dynamic_cast<strided_dataset<T, int64_t>*>(dataset_.get());
+    if (p != nullptr) { return p->view(); }
+    auto d = dataset_->dim();
+    return raft::make_device_strided_matrix_view<const T, int64_t>(nullptr, 0, d, d);
+  }
+
+  /** Dataset [size, dim] */
+  [[nodiscard]] inline auto data() const noexcept -> const cuvs::neighbors::dataset<int64_t>&
+  {
+    return *dataset_;
+  }
+
+  /** vamana graph [size, graph-degree] */
+  [[nodiscard]] inline auto graph() const noexcept
+    -> raft::device_matrix_view<const IdxT, int64_t, raft::row_major>
+  {
+    return graph_view_;
+  }
+
+  // Don't allow copying the index for performance reasons (try avoiding copying data)
+  index(const index&)                    = delete;
+  index(index&&)                         = default;
+  auto operator=(const index&) -> index& = delete;
+  auto operator=(index&&) -> index&      = default;
+  ~index()                               = default;
+
+  /** Construct an empty index. */
+  index(raft::resources const& res,
+        cuvs::distance::DistanceType metric = cuvs::distance::DistanceType::L2Expanded)
+    : cuvs::neighbors::index(),
+      metric_(metric),
+      graph_(raft::make_device_matrix<IdxT, int64_t>(res, 0, 0)),
+      dataset_(new cuvs::neighbors::empty_dataset<int64_t>(0))
+  {
+  }
+
+/** Construct an index from dataset and vamana graph
+   *
+   */
+  template<typename data_accessor, typename graph_accessor>
+  index(raft::resources const& res,
+        cuvs::distance::DistanceType metric,
+        raft::mdspan<const T, raft::matrix_extent<int64_t>, raft::row_major, data_accessor> dataset,
+        raft::mdspan<const IdxT, raft::matrix_extent<int64_t>, raft::row_major, graph_accessor>
+          vamana_graph)
+    : cuvs::neighbors::index(),
+      metric_(metric),
+      graph_(raft::make_device_matrix<IdxT, int64_t>(res, 0, 0)),
+      dataset_(make_aligned_dataset(res, dataset, 16))
+  {
+    RAFT_EXPECTS(dataset.extent(0) == vamana_graph.extent(0),
+                 "Dataset and vamana_graph must have equal number of rows");
+    update_graph(res, vamana_graph);
+
+    raft::resource::sync_stream(res);
+  }
+
+
+  /**
+   * Replace the dataset with a new dataset.
+   *
+   * If the new dataset rows are aligned on 16 bytes, then only a reference is stored to the
+   * dataset. It is the caller's responsibility to ensure that dataset stays alive as long as the
+   * index.
+   */
+  void update_dataset(raft::resources const& res,
```

Comment:
CAGRA construction uses `update_graph` and `update_dataset`, so you will have to refactor a bit or make the private, right? 

Agreed that we can remove for vamana, though, if we don't want it as an exposed API call.

Reply:
Removed `update_dataset`, but left `update_graph` in for now, since build calls it to copy the vamana graph into the index (similar to what cagra does).